docker/sds-api/spine-directory-service/sds/request/cpm.py

```diff
@@ -10,11 +18,139 @@ async def get_device_from_cpm(ods_code: str, interaction_id: str, manufacturing_
         }
     ]
 
+
 async def get_endpoint_from_cpm(ods_code: str, interaction_id: str = None, party_key: str = None) -> List:
     return [
         {
             "ods-code": ods_code,
             "interaction_id": interaction_id,
             "party_key": party_key
         }
-    ]
\ No newline at end of file
+    ]
+
+def process_cpm_endpoint_request(data: dict, query_parts: dict):
+    endpoints = EndpointCpm(data=data, query_parts=query_parts)
+    filtered_endpoints = endpoints.filter_cpm_response()
+    return endpoints.transform_to_ldap(filtered_endpoints)
+
+def process_cpm_device_request(data: dict, query_parts: dict):
+    devices = DeviceCpm(data=data, query_parts=query_parts)
+    filtered_devices = devices.filter_cpm_response()
+    return devices.transform_to_ldap(filtered_devices)
+
+class BaseCpm:
+    FILTER_MAP = {}
+    DATA_MAP = {}
+    DEFAULT_DICT = {}
+    
+    def __init__(self, data, query_parts):
+        self.data = data
+        self.query_parts = query_parts
+
+    def filter_cpm_response(self):
+        filtered_results = []
+        filters = {key: False for key in self.query_parts}
+        
+        for result in self.data["entry"]:
+            for index, res in enumerate(result["entry"]) if "resourceType" in result and result["resourceType"] == "Bundle" else []:
+                for service in res["item"] if "resourceType" in res and res["resourceType"] == "QuestionnaireResponse" else []:
+                    filters = self._check_each_item(self, filters, service)
+                
+            all_filters_true = all(value for value in filters.values())
+            if all_filters_true:
+                filtered_results.append(result["entry"])
+            filters = {key: False for key in filters}
+        
+        return filtered_results
+    
+    @staticmethod
+    def _check_each_item(self, filters, service):
+        for key, value in self.query_parts.items():
+            if service["text"] == self.FILTER_MAP[key]:
+                filters[key] = self._check_match(self, service["answer"], value)
+        return filters
+    
+    @staticmethod
+    def _check_match(self, answers, match):
+        for answer in answers:
+            if answer["valueString"] == match:
+                return True
+    
+    @staticmethod
+    def process_questionnaire_response(self, item, data_dict, ldap_data_mapping):
+        if "resourceType" in item and item["resourceType"] == "QuestionnaireResponse":      
```

Comment:
I think you're being pedantic about the naming aren't you? :rolling_on_the_floor_laughing: My bad, there used to be a  method called process_device_response. I'm assuming that's what you're getting at

Reply:
oh no it wasn't about the naming - mean does this line ever == False?

Reply:
I've actually changed it now anyway, to `if item.get("resourceType") == "QuestionnaireResponse":` inside the transform_to_ldap, but to answer your question, according to the mock, yes it can evaluate to false. because Device as a resourceType actually sits within another layer "resource"